firebase-installations/src/main/java/com/google/firebase/installations/FirebaseInstallations.java

```diff
@@ -103,4 +125,129 @@ String getApplicationId() {
+   *                                     |
+   *                Register the Fid on FIS Server and update the PersistedFidEntry
+   *                                     |
+   *                            return the Persisted Fid
+   *
+   * </pre>
+   */
+  @WorkerThread
+  private String getFirebaseInstallationId() throws FirebaseInstallationsException {
+
+    PersistedFidEntry persistedFidEntry = persistedFid.readPersistedFidEntryValue();
+
+    if (persistedFidExistsAndRegistered(persistedFidEntry)) {
+      return persistedFidEntry.getFirebaseInstallationId();
+    }
+
+    if (persistedFidMissingOrInErrorState(persistedFidEntry)) {
+      return createFidAndPersistFidEntry();
+    }
+
+    if (persistedFidExistsAndUnregistered(persistedFidEntry)) {
+      Tasks.call(executor, () -> registerAndSaveFid(persistedFidEntry.getFirebaseInstallationId()));
+      return persistedFidEntry.getFirebaseInstallationId();
+    }
+
+    return null;
+  }
+
+  private static boolean persistedFidMissingOrInErrorState(PersistedFidEntry persistedFidEntry) {
+    return persistedFidEntry == null
+        || persistedFidEntry.getPersistedStatus() == PersistedFid.PersistedStatus.REGISTER_ERROR;
+  }
+
+  private static boolean persistedFidExistsAndRegistered(PersistedFidEntry persistedFidEntry) {
+    return persistedFidEntry != null
+        && persistedFidEntry.getPersistedStatus() == PersistedFid.PersistedStatus.REGISTERED;
+  }
+
+  private static boolean persistedFidExistsAndUnregistered(PersistedFidEntry persistedFidEntry) {
+    return persistedFidEntry != null
+        && persistedFidEntry.getPersistedStatus() == PersistedFid.PersistedStatus.UNREGISTERED;
+  }
+
+  private String createFidAndPersistFidEntry() throws FirebaseInstallationsException {
+    String fid = Utils.createRandomFid();
+
+    boolean firstUpdateCacheResult =
+        persistedFid.insertOrUpdatePersistedFidEntry(
+            PersistedFidEntry.builder()
+                .setFirebaseInstallationId(fid)
+                .setPersistedStatus(PersistedFid.PersistedStatus.UNREGISTERED)
+                .build());
+
+    if (!firstUpdateCacheResult) {
+      throw new FirebaseInstallationsException(
+          "Failed to update client side cache.",
+          FirebaseInstallationsException.Status.CLIENT_ERROR);
+    }
+
+    Tasks.call(executor, () -> registerAndSaveFid(fid));
+    return fid;
+  }
+
+  /**
+   * Registers the created Fid with FIS Servers if the Network is available and update the shared
+   * prefs.
+   */
+  private Void registerAndSaveFid(String fid) throws FirebaseInstallationsException {
+    if (!Utils.isNetworkAvailable(firebaseApp.getApplicationContext())) {
+      return null;
+    }
+
+    try {
+      long creationTime = Utils.getCurrentTimeInSeconds();
+
+      InstallationResponse installationResponse =
+          serviceClient.createFirebaseInstallation(
+              firebaseApp.getOptions().getApiKey(),
```

Comment:
can you add /* paramName= */ style comments here? otherwise it looks really easy to transpose one of the string params into another

Reply:
Done. https://github.com/firebase/firebase-android-sdk/pull/703/commits/f8e75fd4bbf0b83b84fa71bc360501c862e3d89c